lib/SPIRVProducerPass.cpp

```diff
@@ -3220,144 +3262,802 @@ spv::Op SPIRVProducerPass::GetSPIRVBinaryOpcode(Instruction &I) {
+}
+
+SPIRVID
+SPIRVProducerPass::GenerateSubgroupInstruction(CallInst *Call,
+                                               const FunctionInfo &FuncInfo) {
+  SPIRVID RID;
+
+  // requires SPIRV version 1.3 or greater
+  if (SPIRVMinorVersion < 3)
+    llvm_unreachable(
+        "Sub-group support requires SPIR-V version 1.3 or greater");
+
+  static const std::map<Builtins::BuiltinType, spv::BuiltIn> clspvBuiltinMap = {
+      {Builtins::kGetSubGroupSize, spv::BuiltInSubgroupSize},
+      {Builtins::kGetMaxSubGroupSize, spv::BuiltInSubgroupMaxSize},
+      {Builtins::kGetNumSubGroups, spv::BuiltInNumSubgroups},
+      {Builtins::kGetSubGroupId, spv::BuiltInSubgroupId},
+      {Builtins::kGetSubGroupLocalId, spv::BuiltInSubgroupLocalInvocationId},
+  };
+
+  auto FuncType = FuncInfo.getType();
+
+  auto MII = clspvBuiltinMap.find(FuncType);
+  if (MII != clspvBuiltinMap.end()) {
+    SPIRVOperandVec Ops;
+    Ops << Call->getType() << getSPIRVBuiltin(MII->second);
+
+    return addSPIRVInst(spv::OpLoad, Ops);
+  }
+
+  spv::Capability capability = spv::CapabilityGroupNonUniformBallot;
+  spv::Op op = spv::OpNop;
+  switch (FuncInfo.getType()) {
+  case Builtins::kGetEnqueuedNumSubGroups:
+    break;
+  case Builtins::kSubGroupBarrier:
+    break;
+
+  case Builtins::kSubGroupAll:
+    capability = spv::CapabilityGroupNonUniformVote;
+    op = spv::OpGroupNonUniformAll;
+    break;
+  case Builtins::kSubGroupAny:
+    capability = spv::CapabilityGroupNonUniformVote;
+    op = spv::OpGroupNonUniformAny;
+    break;
+  case Builtins::kSubGroupBroadcast:
+    capability = spv::CapabilityGroupNonUniformBallot;
+    op = spv::OpGroupNonUniformBroadcast;
+    break;
+  case Builtins::kSubGroupReduceAdd:
+  case Builtins::kSubGroupScanExclusiveAdd:
+  case Builtins::kSubGroupScanInclusiveAdd: {
+    capability = spv::CapabilityGroupNonUniformArithmetic;
+    if (FuncInfo.getParameter(0).type_id == Type::IntegerTyID) {
+      op = spv::OpGroupNonUniformIAdd;
+    } else {
+      op = spv::OpGroupNonUniformFAdd;
+    }
+    break;
+  }
+  case Builtins::kSubGroupReduceMin:
+  case Builtins::kSubGroupScanExclusiveMin:
+  case Builtins::kSubGroupScanInclusiveMin: {
+    capability = spv::CapabilityGroupNonUniformArithmetic;
+    auto &param = FuncInfo.getParameter(0);
+    if (param.type_id == Type::IntegerTyID) {
+      op = param.is_signed ? spv::OpGroupNonUniformSMin
+                           : spv::OpGroupNonUniformUMin;
+    } else {
+      op = spv::OpGroupNonUniformFMin;
+    }
+    break;
+  }
+  case Builtins::kSubGroupReduceMax:
+  case Builtins::kSubGroupScanExclusiveMax:
+  case Builtins::kSubGroupScanInclusiveMax: {
+    capability = spv::CapabilityGroupNonUniformArithmetic;
+    auto &param = FuncInfo.getParameter(0);
+    if (param.type_id == Type::IntegerTyID) {
+      op = param.is_signed ? spv::OpGroupNonUniformSMax
+                           : spv::OpGroupNonUniformUMax;
+    } else {
+      // TODO(sjw): what about Half, Double
```

Comment:
This is the correct opcode, but Vulkan requires extra feature bits from subgroup extended types extension to enable support.